ambry-commons/src/main/java/com/github/ambry/commons/RetryExecutor.java

```diff
@@ -67,8 +67,15 @@ public RetryExecutor(ScheduledExecutorService scheduler) {
           int waitTimeMs = policy.waitTimeMs(currAttempts);
           logger.info("{} of {} attempts failed, will keep retrying after a {} ms backoff. exception='{}'",
               currAttempts, policy.maxAttempts(), waitTimeMs, exception);
-          scheduler.schedule(() -> recursiveAsyncRetry(call, isRetriable, policy, userCallback, currAttempts), waitTimeMs,
-              TimeUnit.MILLISECONDS);
+          if (scheduler == null) {
+            if (waitTimeMs > 0) {
+              logger.warn("No scheduler to delay execution of retry by {} ms", waitTimeMs);
+            }
+            recursiveAsyncRetry(call, isRetriable, policy, userCallback, currAttempts);
```

Comment:
If no scheduler, we can still use Thread.sleep() to sleep. Why call recursiveAsyncRetry immediately?